src/query/ee/src/storages/fuse/operations/vacuum_temporary_files.rs

```diff
@@ -34,229 +36,228 @@ const DEFAULT_RETAIN_DURATION: Duration = Duration::from_secs(60 * 60 * 24 * 3);
+    match options {
+        VacuumTempOptions::QueryHook(nodes_num, query_id) => {
+            vacuum_query_hook(abort_checker, &temporary_dir, *nodes_num, query_id, limit).await
+        }
+        VacuumTempOptions::VacuumCommand(duration) => {
+            vacuum_by_duration(abort_checker, &temporary_dir, limit, duration).await
+        }
+    }
+}
+
+async fn vacuum_by_duration(
+    abort_checker: AbortChecker,
+    temporary_dir: &str,
+    mut limit: usize,
+    duration: &Option<Duration>,
+) -> Result<usize> {
+    let operator = DataOperator::instance().operator();
+    let start_time = Instant::now();
+
+    let expire_time = duration.unwrap_or(DEFAULT_RETAIN_DURATION).as_millis() as i64;
     let timestamp = SystemTime::now()
         .duration_since(UNIX_EPOCH)
         .unwrap()
         .as_millis() as i64;
 
-    let operator = DataOperator::instance().operator();
-
+    let mut removed_total = 0;
     let temporary_dir = format!("{}/", temporary_dir.trim_end_matches('/'));
+    let mut ds = operator.lister_with(&temporary_dir).await?;
+
+    let mut temp_files = Vec::new();
+    let mut gc_metas = HashSet::new();
+
+    // We may delete next entries during iteration
+    // So we can't use
+    loop {
+        let de = ds.try_next().await;
+        match de {
+            Ok(Some(de)) => {
+                abort_checker.try_check_aborting()?;
+                if de.path() == temporary_dir {
+                    continue;
+                }
+                let name = de.name();
 
-    let mut ds = operator
-        .lister_with(&temporary_dir)
-        .metakey(Metakey::Mode | Metakey::LastModified)
-        .await?;
-
-    let mut removed_temp_files = 0;
-    let mut total_cleaned_size = 0;
-    let mut total_batch_size = 0;
-    let start_time = Instant::now();
-
-    while removed_temp_files < limit {
-        let instant = Instant::now();
-        let mut end_of_stream = true;
-        let mut remove_temp_files_path = Vec::with_capacity(1000);
-        let mut batch_size = 0;
-
-        while let Some(de) = ds.try_next().await? {
-            abort_checker.try_check_aborting()?;
-            if de.path() == temporary_dir {
-                continue;
-            }
-
-            let meta = de.metadata();
-
-            match meta.mode() {
-                EntryMode::DIR => {
-                    let life_mills =
-                        match operator.exists(&format!("{}finished", de.path())).await? {
-                            true => 0,
-                            false => expire_time,
-                        };
-
-                    vacuum_finished_query(
-                        &abort_checker,
-                        start_time,
-                        &mut removed_temp_files,
-                        &mut total_cleaned_size,
-                        &mut batch_size,
-                        &de,
+                let meta = operator.stat(de.path()).await;
```

Comment:
We can use `de.metadata().mode()` to check if it's a dir.

Reply:
we need to check `last_modified`, so `stat` op can't be avoided.

Reply:
> we need to check `last_modified`, so `stat` op can't be avoided.

`de.metadata()` will also have `last_modified` for most object storage services.